design-proposals/klip-42-schema-migrations-tool.md

```diff
@@ -0,0 +1,442 @@
+| LATEST      | 2       | Add users     | Undone   | <MD5-sum>  | 12-03-2020 10:34:30 | 12-03-2020 10:34:34 | 1        |
++-------------+---------+---------------+----------+------------+---------------------+---------------------+----------+
+```
+
+The tool will later use pull queries to identify the current state of the system:
+```sql
+SELECT version, state, previous FROM SCHEMA_VERSION WHERE version_key = 'CURRENT'; 
+```
+
+
+### Naming convention
+
+Hope you have noticed the naming rules I followed in the previous examples for naming the migration files. Having a naming convention for files is necessary
+for the tool to detect what migration to apply or revert. Naming files are easier than creating configurations or command parameters to specify the same info.
+
+The migration files follow the same `Flyway` naming convention.
+`(Prefix)(Version)__(Description).sql`
+
+The `Prefix` specifies whether the file is a new migration (`V`) file. The prefix is used so that we can add other operations in the future, such as undo operations.
+The `Version` is the version number used for the schema. Versions will not support decimal versions. Integers with 6 digits are used.
+The `Description` is a name or description of the new migration. Description uses underscores (automatically replaced by spaces at runtime) or spaces separated the words.
+
+i.e.
+```
+- V000001__Initial_setup.sql    # a new version to migrate (v1) with name 'Initial setup'
+- V000002__Add_users.sql        # a new version to migrate (v2) with name 'Add users'
+```
+
+It is recommended the user creates these two files on any new migration. The tool will not enforce that. We need specify this recommendation in the ksqlDB documents.
+
+### Dry-runs
+
+A dry-run for ksqlDB migrations will only verify what migrations will be applied in the cluster. It will not attempt to execute or simulate any migration statement found in the files. This
+feature will allow users to test that migration files have the right version and description names as well as know what migrations will be applied in a determined cluster.
+
+### Directory structure
+
+The migrations tool will use following directory structure:
+
+<migrations-project-dir>
+|
+|- ksql-migrations.properties
+|- migrations/
+
+When the migrations tool is executed, it will look at the `migrations` directory (by default) for SQL migration files to execute. This directory can be modified in the  
+configuration file or command line parameters.
+
+If the `ksql-migrations.properties` exist in the root directory, then it will use the configuration provided by the file.
+
+### Command Syntax
+
+Finally, let's look at the rest of the command parameters that will exist to facilitate schema migrations.
+
+```shell script
+Usage:
+  ksql-migrations [options] commands
+  
+Commands
+  new  <project-path>  Creates a new migrations project, directory structure and config file.
+
+  initialize   Initializes the schema version table
```

Comment:
what happens if the user tries to initialize twice? I assume we should detect and fail?

Reply:
Yes. Although, you know. I think this is duplicated. The intention of `initialize` is to create an empty stream and table schema. And the intention of `baseline` is to create the stream and table schema too but with a version specified in the baseline parameter. Both are similar. 

I will remove the `baseline` and add the parameters to `initialize` to initialize to a specified version if desired.